langchain/src/chat_models/fireworks.ts

```diff
@@ -0,0 +1,131 @@
+ *
+ * To use, you should have the `openai` package installed and
+ * the `FIREWORKS_API_KEY` environment variable set.
+ */
+export class ChatFireworks extends ChatOpenAI {
+  static lc_name() {
+    return "ChatFireworks";
+  }
+
+  _llmType() {
+    return "fireworks";
+  }
+
+  get lc_secrets(): { [key: string]: string } | undefined {
+    return {
+      fireworksApiKey: "FIREWORKS_API_KEY",
+    };
+  }
+
+  lc_serializable = true;
+
+  fireworksApiKey?: string;
+
+  constructor(
+    fields?: Partial<
+      Omit<OpenAIChatInput, "openAIApiKey" | FireworksUnsupportedArgs>
+    > &
+      BaseChatModelParams & { fireworksApiKey?: string }
+  ) {
+    const fireworksApiKey =
+      fields?.fireworksApiKey || getEnvironmentVariable("FIREWORKS_API_KEY");
+
+    if (!fireworksApiKey) {
+      throw new Error(
+        `Fireworks API key not found. Please set the FIREWORKS_API_KEY environment variable or provide the key into "fireworksApiKey"`
+      );
+    }
+
+    super({
+      ...fields,
+      modelName:
+        fields?.modelName || "accounts/fireworks/models/llama-v2-13b-chat",
+      openAIApiKey: fireworksApiKey,
+      configuration: {
+        baseURL: "https://api.fireworks.ai/inference/v1",
+      },
+    });
+
+    this.fireworksApiKey = fireworksApiKey;
+  }
+
+  toJSON() {
+    const result = super.toJSON();
+
+    if (
+      "kwargs" in result &&
+      typeof result.kwargs === "object" &&
+      result.kwargs != null
+    ) {
+      delete result.kwargs.openai_api_key;
+      delete result.kwargs.configuration;
+    }
+
+    return result;
+  }
+
+  async completionWithRetry(
+    request: OpenAIClient.Chat.ChatCompletionCreateParamsStreaming,
+    options?: OpenAICoreRequestOptions
+  ): Promise<AsyncIterable<OpenAIClient.Chat.Completions.ChatCompletionChunk>>;
+
+  async completionWithRetry(
+    request: OpenAIClient.Chat.ChatCompletionCreateParamsNonStreaming,
+    options?: OpenAICoreRequestOptions
+  ): Promise<OpenAIClient.Chat.Completions.ChatCompletion>;
+
+  /**
+   * Calls the Fireworks API with retry logic in case of failures.
+   * @param request The request to send to the Fireworks API.
+   * @param options Optional configuration for the API call.
+   * @returns The response from the Fireworks API.
+   */
+  async completionWithRetry(
+    request:
+      | OpenAIClient.Chat.ChatCompletionCreateParamsStreaming
+      | OpenAIClient.Chat.ChatCompletionCreateParamsNonStreaming,
+    options?: OpenAICoreRequestOptions
+  ): Promise<
+    | AsyncIterable<OpenAIClient.Chat.Completions.ChatCompletionChunk>
+    | OpenAIClient.Chat.Completions.ChatCompletion
+  > {
+    delete request.frequency_penalty;
+    delete request.presence_penalty;
+    delete request.logit_bias;
+    delete request.functions;
+
+    if (request.stream === true) {
+      return super.completionWithRetry(request, options);
+    }
+
+    return super.completionWithRetry(request, options);
+  }
+
+  bind(
```

Comment:
This is fine but should we just add a `CallOption` type arg to OpenAI/ChatOpenAI instead?

Reply:
That's a possibility, but I feel kind of weird modifying the OpenAI class to support other LLMs, wdyt?